dotCMS/src/main/java/com/dotcms/ai/api/CompletionsAPIImpl.java

```diff
@@ -0,0 +1,299 @@
+                    WebAPILocator.getHostWebAPI().getCurrentHostNoThrow(HttpServletRequestThreadLocal.INSTANCE.getRequest()))
+            .getOrElse(APILocator.systemHost()))
+    );
+
+
+    public CompletionsAPIImpl() {
+        this(null);
+    }
+
+
+    public CompletionsAPIImpl(Lazy<AppConfig> config) {
+
+        this.config = (config != null)
+                ? config
+                : defaultConfig;
+    }
+
+    @Override
+    public JSONObject summarize(CompletionsForm summaryRequest) {
+        EmbeddingsDTO searcher = EmbeddingsDTO.from(summaryRequest).build();
+
+        List<EmbeddingsDTO> localResults = EmbeddingsAPI.impl().getEmbeddingResults(searcher);
+
+        // send all this as a json blob to OpenAI
+        JSONObject json = buildRequestJson(summaryRequest, localResults);
+        if (json.optBoolean("stream", false)) {
+            throw new DotRuntimeException("Please use the summarizeStream method to stream results");
+        }
+        json.put("stream", false);
+
+        String openAiResponse = Try.of(() -> OpenAIRequest.doRequest(config.get().getApiUrl(), "post", config.get().getApiKey(), json)).getOrElseThrow(DotRuntimeException::new);
+
+        JSONObject dotCMSResponse = EmbeddingsAPI.impl().reduceChunksToContent(searcher, localResults);
+
+        dotCMSResponse.put("openAiResponse", new JSONObject(openAiResponse));
+        return dotCMSResponse;
+    }
+
+    private JSONObject buildRequestJson(CompletionsForm form, List<EmbeddingsDTO> searchResults) {
+
+
+        OpenAIModel model = OpenAIModel.resolveModel(form.model);
+        // aggregate matching results into text
+        final StringBuilder supportingContent = new StringBuilder();
+        searchResults.forEach(s -> supportingContent.append(s.extractedText).append(" "));
+
+
+        final String systemPrompt = getSystemPrompt(form.prompt, supportingContent.toString());
+        String textPrompt = getTextPrompt(form.prompt, supportingContent.toString());
+
+        final int systemPromptTokens = countTokens(systemPrompt);
+        textPrompt = reduceStringToTokenSize(textPrompt, model.maxTokens - form.responseLengthTokens - systemPromptTokens);
+
+
+        final JSONObject json = new JSONObject();
+
+        json.put("stream", form.stream);
+        json.put("temperature", form.temperature);
+        final List<Map> messages = new ArrayList<>();
+        if (UtilMethods.isSet(systemPrompt)) {
+            messages.add(Map.of("role", "system", "content", systemPrompt));
+        }
+        messages.add(Map.of("role", "user", "content", textPrompt));
+        json.put("messages", messages);
+
+        if (UtilMethods.isSet(form.model)) {
+            json.put("model", model.modelName);
+        }
+
+        if (form.responseLengthTokens > 0) {
+            json.put("max_tokens", form.responseLengthTokens);
+        }
+
+
+        return json;
+    }
+
+    private String getSystemPrompt(String prompt, String supportingContent) {
```

Comment:
Aren't these two methods really similar and probably most of its logic can be moved to common method?

Reply:
Doing this on the following PR